mypy reports issues in the `tests` directory

If you modify `mypy.ini` so that the `files = manticore` line is changed to `files = manticore, tests`, then mypy reports a number of issues within the `tests` directory:
```
tests/other/utils/test_config.py:130: error: Name 'test_parse' already defined on line 113
tests/ethereum/test_detectors.py:92: error: Incompatible types in assignment (expression has type "Type[DetectUnusedRetVal]", base class "EthDetectorTest" defined the type as "None")
tests/ethereum/test_detectors.py:112: error: Incompatible types in assignment (expression has type "Type[DetectSuicidal]", base class "EthDetectorTest" defined the type as "None")
tests/ethereum/test_detectors.py:134: error: Incompatible types in assignment (expression has type "Type[DetectExternalCallAndLeak]", base class "EthDetectorTest" defined the type as "None")
tests/ethereum/test_detectors.py:242: error: Incompatible types in assignment (expression has type "Type[DetectEnvInstruction]", base class "EthDetectorTest" defined the type as "None")
tests/ethereum/test_detectors.py:265: error: Incompatible types in assignment (expression has type "Type[DetectDelegatecall]", base class "EthDetectorTest" defined the type as "None")
tests/ethereum/test_detectors.py:304: error: Incompatible types in assignment (expression has type "Type[DetectRaceCondition]", base class "EthDetectorTest" defined the type as "None")
tests/ethereum/test_detectors.py:403: error: Incompatible types in assignment (expression has type "Type[DetectManipulableBalance]", base class "EthDetectorTest" defined the type as "None")
tests/native/test_state.py:213: error: Name 'testContextSerialization' already defined on line 160
tests/native/test_memory.py:1431: error: Not all arguments converted during string formatting
tests/native/test_cpu_manual.py:11: error: Incompatible import of "Memory" (imported name has type "Type[tests.native.mockmem.Memory]", local name has type "Type[manticore.native.memory.Memory]")
tests/native/test_cpu_automatic.py:6: error: Cannot resolve name "solver" (possible cyclic definition)
Found 12 errors in 6 files (checked 171 source files)
```

--- a/manticore/utils/log.py
+++ b/manticore/utils/log.py
@@ -1,9 +1,11 @@
 import logging
 import sys
 
+from typing import Set
+
 manticore_verbosity = 0
 DEFAULT_LOG_LEVEL = logging.WARNING
-all_loggers = set()
+all_loggers: Set[str] = set()
 default_factory = logging.getLogRecordFactory()
 logfmt = "%(asctime)s: [%(process)d] %(name)s:%(levelname)s %(message)s"
 handler = logging.StreamHandler(sys.stdout)

--- a/manticore/utils/log.py
+++ b/manticore/utils/log.py
@@ -1,7 +1,7 @@
 import logging
 import sys
 
-from typing import Set
+from typing import List, Set, Tuple
 
 manticore_verbosity = 0
 DEFAULT_LOG_LEVEL = logging.WARNING
@@ -18,7 +18,7 @@ class ContextFilter(logging.Filter):
     This is a filter which injects contextual information into the log.
     """
 
-    def summarized_name(self, name):
+    def summarized_name(self, name: str) -> str:
         """
         Produce a summarized record name
           i.e. manticore.core.executor -> m.c.executor
@@ -44,7 +44,7 @@ class ContextFilter(logging.Filter):
     colored_levelname_format = "\x1b[{}m{}:\x1b[0m"
     plain_levelname_format = "{}:"
 
-    def colored_level_name(self, levelname):
+    def colored_level_name(self, levelname: str) -> str:
         """
         Colors the logging level in the logging record
         """
@@ -53,7 +53,7 @@ class ContextFilter(logging.Filter):
         else:
             return self.colored_levelname_format.format(self.color_map[levelname], levelname)
 
-    def filter(self, record):
+    def filter(self, record) -> bool:
         record.name = self.summarized_name(record.name)
         record.levelname = self.colored_level_name(record.levelname)
         return True
@@ -67,7 +67,7 @@ class CustomLogger(logging.Logger):
     Custom Logger class that can grab the correct verbosity level from this module
     """
 
-    def __init__(self, name, level=DEFAULT_LOG_LEVEL, *args):
+    def __init__(self, name: str, level=DEFAULT_LOG_LEVEL, *args) -> None:
         super().__init__(name, min(get_verbosity(name), level), *args)
         all_loggers.add(name)
         self.initialized = False
@@ -81,11 +81,11 @@ class CustomLogger(logging.Logger):
 logging.setLoggerClass(CustomLogger)
 
 
-def disable_colors():
+def disable_colors() -> None:
     ContextFilter.colors_disabled = True
 
 
-def get_levels():
+def get_levels() -> List[List[Tuple[str, int]]]:
     return [
         # 0
         [(x, DEFAULT_LOG_LEVEL) for x in all_loggers],
@@ -125,8 +125,8 @@ def get_levels():
     ]
 
 
-def get_verbosity(logger_name):
-    def match(name, pattern):
+def get_verbosity(logger_name: str) -> int:
+    def match(name: str, pattern: str):
         """
         Pseudo globbing that only supports full fields. 'a.*.d' matches 'a.b.d'
         but not 'a.b.c.d'.
@@ -148,7 +148,7 @@ def get_verbosity(logger_name):
     return DEFAULT_LOG_LEVEL
 
 
-def set_verbosity(setting):
+def set_verbosity(setting: int) -> None:
     global manticore_verbosity
     manticore_verbosity = min(max(setting, 0), len(get_levels()) - 1)
     for logger_name in all_loggers:

--- a/mypy.ini
+++ b/mypy.ini
@@ -1,6 +1,6 @@
 [mypy]
 python_version = 3.6
-files = manticore
+files = manticore, tests
 
 # Generated file
 [mypy-manticore.ethereum.parsetab]

--- a/tests/ethereum/test_detectors.py
+++ b/tests/ethereum/test_detectors.py
@@ -16,6 +16,7 @@ from manticore.ethereum import (
     DetectExternalCallAndLeak,
     DetectIntegerOverflow,
     DetectManipulableBalance,
+    Detector,
     DetectRaceCondition,
     DetectSuicidal,
     DetectUnusedRetVal,
@@ -25,6 +26,8 @@ from manticore.ethereum import (
 from manticore.ethereum.plugins import LoopDepthLimiter, KeepOnlyIfStorageChanges
 
 from manticore.utils import config, log
+
+from typing import Type
 
 consts = config.get_group("core")
 consts.mprocessing = consts.mprocessing.single
@@ -39,11 +42,8 @@ def make_mock_evm_state():
 
 
 class EthDetectorTest(unittest.TestCase):
-    """
-    Subclasses must assign this class variable to the class for the detector
-    """
-
-    DETECTOR_CLASS = None
+    # Subclasses must assign this class variable to the class for the detector
+    DETECTOR_CLASS: Type[Detector]
 
     def setUp(self):
         self.mevm = ManticoreEVM()

--- a/tests/ethereum/test_detectors.py
+++ b/tests/ethereum/test_detectors.py
@@ -11,15 +11,15 @@ from manticore.platforms.evm import EVMWorld
 
 from manticore.core.smtlib import operators, ConstraintSet
 from manticore.ethereum import (
-    ManticoreEVM,
-    DetectIntegerOverflow,
-    DetectUnusedRetVal,
-    DetectSuicidal,
     DetectDelegatecall,
-    DetectExternalCallAndLeak,
     DetectEnvInstruction,
-    DetectRaceCondition,
+    DetectExternalCallAndLeak,
+    DetectIntegerOverflow,
     DetectManipulableBalance,
+    DetectRaceCondition,
+    DetectSuicidal,
+    DetectUnusedRetVal,
+    ManticoreEVM,
     State,
 )
 from manticore.ethereum.plugins import LoopDepthLimiter, KeepOnlyIfStorageChanges

--- a/tests/ethereum/test_sha3.py
+++ b/tests/ethereum/test_sha3.py
@@ -20,12 +20,6 @@ THIS_DIR = os.path.dirname(os.path.abspath(__file__))
 
 
 class EthSha3TestSymbolicate(unittest.TestCase):
-    """
-    Subclasses must assign this class variable to the class for the detector
-    """
-
-    DETECTOR_CLASS = None
-
     def setUp(self):
         evm_consts = config.get_group("evm")
         evm_consts.sha3 = evm_consts.sha3.symbolicate
@@ -426,12 +420,6 @@ keccak256(keccak256(keccak256(keccak256(keccak256(keccak256(keccak256(keccak256(
 
 
 class EthSha3TestConcrete(unittest.TestCase):
-    """
-    Subclasses must assign this class variable to the class for the detector
-    """
-
-    DETECTOR_CLASS = None
-
     def setUp(self):
         evm_consts = config.get_group("evm")
         evm_consts.sha3 = evm_consts.sha3.concretize

--- a/tests/native/test_cpu_automatic.py
+++ b/tests/native/test_cpu_automatic.py
@@ -1,9 +1,9 @@
 import unittest
 from manticore.native.cpu.x86 import *
-from manticore.core.smtlib import *
+import manticore.core.smtlib
 from manticore.native.memory import *
 
-solver = solver.Z3Solver.instance()
+solver = manticore.core.smtlib.solver.Z3Solver.instance()
 
 
 class CPUTest(unittest.TestCase):

--- a/tests/native/test_cpu_manual.py
+++ b/tests/native/test_cpu_manual.py
@@ -8,7 +8,7 @@ from manticore.native.cpu.x86 import AMD64Cpu
 from manticore.native.memory import *
 from manticore.core.smtlib import BitVecOr, operator, Bool
 from manticore.core.smtlib.solver import Z3Solver
-from .mockmem import Memory
+from .mockmem import Memory as MockMemory
 from functools import reduce
 
 solver = Z3Solver.instance()
@@ -186,7 +186,7 @@ class SymCPUTest(unittest.TestCase):
             return self.value
 
     def setUp(self):
-        mem = Memory()
+        mem = MockMemory()
         self.cpu = I386Cpu(mem)  # TODO reset cpu in between tests...
         # TODO mock getchar/putchar in case the instruction accesses memory directly
 

--- a/tests/native/test_memory.py
+++ b/tests/native/test_memory.py
@@ -1428,7 +1428,7 @@ class MemoryTest(unittest.TestCase):
         # No Access Reading <4160741376>
         # self.assertRaisesRegexp(MemoryException, r"No access reading.*", mem.__getitem__, x)
         with self.assertRaisesRegex(
-            InvalidSymbolicMemoryAccess, "Invalid symbolic memory access.*".format(addr)
+            InvalidSymbolicMemoryAccess, f"Invalid symbolic memory access.*<{addr:x}>"
         ):
             _ = mem[x]
             # mem[addr] = 'a'

--- a/tests/native/test_memory.py
+++ b/tests/native/test_memory.py
@@ -1428,7 +1428,7 @@ class MemoryTest(unittest.TestCase):
         # No Access Reading <4160741376>
         # self.assertRaisesRegexp(MemoryException, r"No access reading.*", mem.__getitem__, x)
         with self.assertRaisesRegex(
-            InvalidSymbolicMemoryAccess, f"Invalid symbolic memory access.*<{addr:x}>"
+            InvalidSymbolicMemoryAccess, "Invalid symbolic memory access \(mode:r\)"
         ):
             _ = mem[x]
             # mem[addr] = 'a'

--- a/tests/native/test_memory.py
+++ b/tests/native/test_memory.py
@@ -1373,7 +1373,7 @@ class MemoryTest(unittest.TestCase):
 
         mem.mprotect(addr, size, "w")
         with self.assertRaisesRegex(
-            InvalidMemoryAccess, f"Invalid memory access \(mode:.\) <{addr:x}>"
+            InvalidMemoryAccess, fr"Invalid memory access \(mode:.\) <{addr:x}>"
         ):
             _ = mem[addr]
 
@@ -1428,7 +1428,7 @@ class MemoryTest(unittest.TestCase):
         # No Access Reading <4160741376>
         # self.assertRaisesRegexp(MemoryException, r"No access reading.*", mem.__getitem__, x)
         with self.assertRaisesRegex(
-            InvalidSymbolicMemoryAccess, "Invalid symbolic memory access \(mode:r\)"
+            InvalidSymbolicMemoryAccess, r"Invalid symbolic memory access \(mode:r\)"
         ):
             _ = mem[x]
             # mem[addr] = 'a'
@@ -1445,7 +1445,7 @@ class MemoryTest(unittest.TestCase):
         mem[addr] = "a"
         mem.mprotect(addr, size, "r")
         with self.assertRaisesRegex(
-            InvalidMemoryAccess, f"Invalid memory access \(mode:w\) <{addr:x}>"
+            InvalidMemoryAccess, fr"Invalid memory access \(mode:w\) <{addr:x}>"
         ):
             mem[addr] = "a"
 
@@ -1461,7 +1461,7 @@ class MemoryTest(unittest.TestCase):
         mem[addr] = "a"
 
         with self.assertRaisesRegex(
-            InvalidMemoryAccess, f"Invalid memory access \(mode:r\) <{addr:x}>"
+            InvalidMemoryAccess, fr"Invalid memory access \(mode:r\) <{addr:x}>"
         ):
             _ = mem[addr]
 

--- a/tests/native/test_state.py
+++ b/tests/native/test_state.py
@@ -210,59 +210,6 @@ class StateTest(unittest.TestCase):
         new_new_state = pickle.loads(new_new_file)
         self.assertEqual(new_new_state.context["step"], 30)
 
-    def testContextSerialization(self):
-        import pickle as pickle
-
-        initial_file = ""
-        new_file = ""
-        new_new_file = ""
-        constraints = ConstraintSet()
-        initial_state = State(constraints, FakePlatform())
-        initial_state.context["step"] = 10
-        initial_file = pickle_dumps(initial_state)
-        with initial_state as new_state:
-            self.assertEqual(initial_state.context["step"], 10)
-            self.assertEqual(new_state.context["step"], 10)
-
-            new_state.context["step"] = 20
-
-            self.assertEqual(initial_state.context["step"], 10)
-            self.assertEqual(new_state.context["step"], 20)
-            new_file = pickle_dumps(new_state)
-
-            with new_state as new_new_state:
-                self.assertEqual(initial_state.context["step"], 10)
-                self.assertEqual(new_state.context["step"], 20)
-                self.assertEqual(new_new_state.context["step"], 20)
-
-                new_new_state.context["step"] += 10
-
-                self.assertEqual(initial_state.context["step"], 10)
-                self.assertEqual(new_state.context["step"], 20)
-                self.assertEqual(new_new_state.context["step"], 30)
-
-                new_new_file = pickle_dumps(new_new_state)
-
-                self.assertEqual(initial_state.context["step"], 10)
-                self.assertEqual(new_state.context["step"], 20)
-                self.assertEqual(new_new_state.context["step"], 30)
-
-            self.assertEqual(initial_state.context["step"], 10)
-            self.assertEqual(new_state.context["step"], 20)
-
-        self.assertEqual(initial_state.context["step"], 10)
-
-        del initial_state
-        del new_state
-        del new_new_state
-
-        initial_state = pickle.loads(initial_file)
-        self.assertEqual(initial_state.context["step"], 10)
-        new_state = pickle.loads(new_file)
-        self.assertEqual(new_state.context["step"], 20)
-        new_new_state = pickle.loads(new_new_file)
-        self.assertEqual(new_new_state.context["step"], 30)
-
 
 class StateMergeTest(unittest.TestCase):
 

--- a/tests/other/utils/test_config.py
+++ b/tests/other/utils/test_config.py
@@ -127,7 +127,7 @@ class ConfigTest(unittest.TestCase):
         self.assertEqual(g.var3, [1, 2, 3])
         self.assertEqual(g.var4, [])
 
-    def test_parse(self):
+    def test_parse_badconfig(self):
         conf = "bad config"
         f = io.StringIO(conf)
         # this shouldn't raise

